crates/cargo-util/src/registry.rs

```diff
@@ -11,11 +11,23 @@ pub fn make_dep_path(dep_name: &str, prefix_only: bool) -> String {
     } else {
         ("/", dep_name)
     };
-    match dep_name.len() {
-        1 => format!("1{}{}", slash, name),
-        2 => format!("2{}{}", slash, name),
-        3 => format!("3/{}{}{}", &dep_name[..1], slash, name),
-        _ => format!("{}/{}{}{}", &dep_name[0..2], &dep_name[2..4], slash, name),
+
+    let mut chars = [None; 4];
```

Comment:
I've tried to optimize the code in that way it shouldn't allocate nor call unwrap, but it seems not so clear as previous one.

Reply:
You're right. Actually the original one (the panicking one) is most readable IMO. And your first attempt also got a better looking than mine and the slice pattern one.

Just made a micro benchmark against top 10 downloads crates on crates.io. The performance difference is statistically significant. The real world difference roughly equals **50ns** times **number of crates in the dependency tree**. Say 200 crates, and that is like **10ms** difference in a cargo build. 

I wouldn't say it's significant for a build. There are so many other places can optimise than this tiny function. I am happy with every solution here if this PR eventually gets approved and merged :)

---

[Criterion benchmark report](https://shade-bittersweet-protocol.glitch.me/report/)
[Benchmark code](https://gist.github.com/weihanglo/2548c8f850be296f3ee39f4266a50dbf)

The result above was tested on:

```
cargo 1.65.0-nightly (efd4ca3dc 2022-08-12)
release: 1.65.0-nightly
commit-hash: efd4ca3dc0b89929dc8c5f5c023d25978d76cb61
commit-date: 2022-08-12
host: aarch64-apple-darwin
libgit2: 1.4.2 (sys:0.14.2 vendored)
libcurl: 7.79.1 (sys:0.4.55+curl-7.83.1 system ssl:(SecureTransport) LibreSSL/3.3.6)
os: Mac OS 12.5.0 [64-bit]
```